src/runner.py

```diff
@@ -88,596 +47,416 @@ class LxdInstanceConfig(TypedDict):
     profiles: List[str]
 
 
-class RunnerInfo:
-    """RunnerInfo stores the info on a runner.
+class Runner:
+    """Single instance of GitHub self-hosted runner.
 
-    Attributes:
-        name (str): Name of the runner.
-        local (pylxd.models.Instance): Local LXD instance information of runner.
-        remote (RemoteRunner): GitHub API information of runner.
+    Attrs:
+        app_name (str): Name of the charm.
+        path (GitHubPath): Path to GitHub repo or org.
+        proxies (ProxySetting): HTTP proxy setting for juju charm.
+        name (str): Name of the runner instance.
+        exist (bool): Whether the runner instance exists on LXD.
+        online (bool): Whether GitHub marks this runner as online.
+        busy (bool): Whether GitHub marks this runner as busy.
     """
 
+    runner_application = Path("/opt/github-runner")
+    env_file = runner_application / ".env"
+    config_script = runner_application / "config.sh"
+    runner_script = runner_application / "start.sh"
+
     def __init__(
-        self, name: str, local: Optional[pylxd.models.Instance], remote: Optional[RemoteRunner]
-    ) -> None:
-        """Construct a instance of runner information.
+        self,
+        github: GhApi,
+        jinja: jinja2.Environment,
+        lxd: pylxd.Client,
+        app_name: str,
+        path: GitHubPath,
+        proxies: ProxySetting,
+        name: str,
+        exist: bool = False,
+        online: bool = False,
+        busy: bool = False,
+        instance: Optional[pylxd.models.Instance] = None,
+    ):
+        """Construct the runner instance.
 
         Args:
+            github: Used to query GitHub API.
+            jinja: Used for working with template.
+            lxd: Used to interact with LXD API.
+            app_name: Name of the charm.
+            path: Path to GitHub repo or org.
+            proxies: HTTP proxy setting for juju charm.
             name: Name of the runner.
-            local: Local LXD instance information of runner.
-            remote: GitHub API information of runner.
+            exist: Whether the runner instance exists on LXD. Defaults to False.
+            online: Whether GitHub marks this runner as online. Defaults to False.
+            busy: Whether GitHub marks this runner as busy. Defaults to False.
+            instance: LXD instance of the runner if already created. Defaults to None.
         """
-        self.name = name
-        self.local = local
-        self.remote = remote
+        # Dependency injection to share the instances across different `Runner` instance.
```

Comment:
Concern over resource usages.

I have not look into the implementation, so it might not be an issue. However, it might be possible that sharing the same instances across multiple runners would allow for reusing TCP connection or websockets, etc. 

Depending on how it is implemented it is still possible for it to share these underlying resource even if we create a `pylxd.Client`, `GhApi`, etc. per runner.